File: content/snippets/go/interview_lcci_03_06_animal_shelter.go
package main

import "fmt"

type Animal struct {
	order int
	id    int
}

type AnimalShelter struct {
	ticket int
	dogs   []Animal
	cats   []Animal
}

func (s *AnimalShelter) Enqueue(id int, kind string) {
	item := Animal{order: s.ticket, id: id}
	s.ticket++
	if kind == "dog" {
		s.dogs = append(s.dogs, item)
	} else {
		s.cats = append(s.cats, item)
	}
}

func (s *AnimalShelter) DequeueAny() int {
	if len(s.dogs) == 0 && len(s.cats) == 0 {
		return -1
	}
	if len(s.dogs) == 0 {
		return s.DequeueCat()
	}
	if len(s.cats) == 0 {
		return s.DequeueDog()
	}
	if s.dogs[0].order < s.cats[0].order {
		return s.DequeueDog()
	}
	return s.DequeueCat()
}

func (s *AnimalShelter) DequeueDog() int {
	if len(s.dogs) == 0 {
		return -1
	}
	id := s.dogs[0].id
	s.dogs = s.dogs[1:]
	return id
}

func (s *AnimalShelter) DequeueCat() int {
	if len(s.cats) == 0 {
		return -1
	}
	id := s.cats[0].id
	s.cats = s.cats[1:]
	return id
}

func main() {
	s := &AnimalShelter{}
	s.Enqueue(1, "dog")
	s.Enqueue(2, "cat")
	s.Enqueue(3, "dog")
	fmt.Println(s.DequeueAny())
	fmt.Println(s.DequeueCat())
	fmt.Println(s.DequeueDog())
	fmt.Println(s.DequeueAny())
}

File: content/snippets/typescript/interview_lcci_03_06_animal_shelter.ts
type Animal = {
  order: number;
  id: number;
};

class AnimalShelter {
  private ticket = 0;
  private dogs: Animal[] = [];
  private cats: Animal[] = [];

  enqueue(id: number, kind: "dog" | "cat"): void {
    const item: Animal = { order: this.ticket, id };
    this.ticket += 1;
    if (kind === "dog") {
      this.dogs.push(item);
    } else {
      this.cats.push(item);
    }
  }

  dequeueAny(): number {
    if (!this.dogs.length && !this.cats.length) {
      return -1;
    }
    if (!this.dogs.length) {
      return this.dequeueCat();
    }
    if (!this.cats.length) {
      return this.dequeueDog();
    }
    return this.dogs[0].order < this.cats[0].order ? this.dequeueDog() : this.dequeueCat();
  }

  dequeueDog(): number {
    return this.dogs.length ? this.dogs.shift()!.id : -1;
  }

  dequeueCat(): number {
    return this.cats.length ? this.cats.shift()!.id : -1;
  }
}

function main(): void {
  const shelter = new AnimalShelter();
  shelter.enqueue(1, "dog");
  shelter.enqueue(2, "cat");
  shelter.enqueue(3, "dog");
  console.log(shelter.dequeueAny());
  console.log(shelter.dequeueCat());
  console.log(shelter.dequeueDog());
  console.log(shelter.dequeueAny());
}

main();
